sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala

```diff
@@ -152,22 +154,28 @@ class DataFrameRangeSuite extends QueryTest with SharedSQLContext with Eventuall
   }
 
   test("Cancelling stage in a query with Range.") {
+    val slices = 10
+
     val listener = new SparkListener {
-      override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
-        eventually(timeout(10.seconds), interval(1.millis)) {
-          assert(DataFrameRangeSuite.stageToKill > 0)
+      override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+        eventually(timeout(10.seconds)) {
+          assert(DataFrameRangeSuite.isTaskStarted)
         }
-        sparkContext.cancelStage(DataFrameRangeSuite.stageToKill)
+        sparkContext.cancelStage(taskStart.stageId)
+        DataFrameRangeSuite.semaphore.release(slices)
```

Comment:
I know this is what the other test does, but after looking at this again I think this is still dangerous.

`sparkContext.cancelStage` is asynchronous, so you might wake up all the tasks and they might finish before the cancellation actually happens. If you just remove this line, then you guarantee the tasks won't finish unless the state is canceled.

Which means you don't really need the semaphore, you can just sleep indefinitely in the tasks (e.g. call `wait()`). And instead of the `eventually` above you could use a `CountDownLatch`.

And if you think of it, you don't need the `DataFrameSuite` object since everything here is local to this test.

All this also applies to the other test, so if you feel like cleaning up that one in a separate PR...

Reply:
The concept is clear.
```
sparkContext.range(0, 10000L, numSlices = slices).mapPartitions { x =>
  synchronized { wait() }
  x
}.toDF("id").agg(sum("id")).collect()
```
throws:
```
Expected the cause to be SparkException, got java.io.NotSerializableException: org.scalatest.Assertions$AssertionsHelper
Serialization stack:
	- object not serializable (class: org.scalatest.Assertions$AssertionsHelper, value: org.scalatest.Assertions$AssertionsHelper@aced190)
...
```
I mean this.

Reply:
The do:

```
val lock = new Object()
lock.synchronized { lock.wait() }
```

Again, you just want to go to sleep waiting for an interrupt. There's like a thousand ways to do that.

Reply:
If we do that so the stopped threads not get died. In the RDD iterator the following code makes sure that the thread will be killed:
```
private[spark] override def killTaskIfInterrupted(): Unit = {
 val reason = reasonIfKilled
  if (reason.isDefined) {
    throw new TaskKilledException(reason.get)
  }
}
```
If `SparkContext.SPARK_JOB_INTERRUPT_ON_CANCEL` will be set then they will die properly.